### ISSUE TITLE
Re-organize API reference docs to be one page per class

### ISSUE THREAD

**Issue by bryevdv** [bokeh/bokeh#13814]
(This issue mostly pertains to the contents of `bokeh.models`)

Our reference docs are currently hand-curated into topical groupings. The original reason for the the groupings was that early on we had issues supporting and effective search capability, and because the API surface was much smaller and easier to organize by hand. Search has since gotten better and the API surface has grown considerably.

In addition to posing a manual maintenance burden, this organization also results in som very heavy pages in our docs. These large page sizes can cause problems or block new features:

* https://github.com/bokeh/bokeh/issues/13799
* https://github.com/bokeh/bokeh/issues/12707

I think we must split the API docs into a one-class-per-page organization. I.e. instead of:

```
https://docs.bokeh.org/en/latest/docs/reference/models/tools.html#bokeh.models.TapTool
```

there would be a dedicated page like:
```
https://docs.bokeh.org/en/latest/docs/reference/models/TapTool
```

This organization is common in other projects and is supported by some tools e.g. [`autoapi`](https://sphinx-autoapi.readthedocs.io/en/latest/) and others. It will take some research and experimentation to see whether these tools can be used "out of the box" or if we will need to make other changes to accommodate them or (worst-case) develop our our tooling. I'll try to do some minimal investigation to see how/if existing tools currently function.

There are some questions that probably intersect with this, such as the idea to split up the monolithic `bokeh.models` namespace itself at the user-facing API level (cc @mattpap I thought there was an issue/discussion about this but I can't find it)

cc @bokeh/dev for comments (not triaging as "discussion" since I think the necessity of change is making itself evident)

**Comment by ianthomas23** [comment]
I am very much in favour of this.

There might be some overlap with BokehJS if we manage to elevate that to a standalone NPM package as we'd need API docs for that, and ideally both sets of API docs would be organised along similar lines for ease of mentally switching between the two. The BokehJS model classes are already split into sensible groups so it would be good to do similar on the Python side.

**Comment by bryevdv** [comment]
@ianthomas23 can you elaborate? The structure on Bokeh sides should still be identical AFAIK. The only difference is that on the Python side model definitions are small so a module like `models/ranges.py` defines several models in one file. But implementations on the JS side are typically much larger, so are split out one more level, with a `models/ranges` *directory*, that contains individual files for each model implementation. 

I'm not sure how that pertains here, though. These things are all distinct:

* API presented to users 
* file/module layout on disk
* organization of the docs

So just being clear, this issues is *not* about expecting users do `from bokeh.models.ranges import Range1d` instead of `from bokeh.models import Range1d`. IMO that is Bokeh 4.0 territory. It's also not about splitting up `ranges.py` to a `ranges` directory to match BokehJS (that seems excessively fine-grained on the Python side) This issue is just about the docs presentation and can be summed up "stop generating enormous pages".

**Comment by bryevdv** [comment]
Just noting that I can get a basic configuration with `autoapi` building (with some new warnings) but for the results are still mixes, e.g. the help strings for properties are not currently being rendered, so some continued investigation is needed.

**Comment by ianthomas23** [comment]
@bryevdv I assumed that

> There are some questions that probably intersect with this, such as the idea to split up the monolithic bokeh.models namespace itself at the user-facing API level (cc @mattpap I thought there was an issue/discussion about this but I can't find it)

meant splitting up the Python directory/class structure to be more fine-grained, but you have said this isn't so. My mistake.

**Comment by bryevdv** [comment]
> meant splitting up the Python directory/class structure to be more fine-grained, but you have said this isn't so. My mistake.

@ianthomas23 no mistake I just hadn't imagined that was something people wanted, since it seems like it will result in hundreds of very small files. I'd like to understand what benefits you identify in splitting things up more are? Entirely possible I just don't see them, but I can learn :)

**Comment by thomktz** [comment]
I think the docs over at [@snehilvj/dash-mantine-components](https://github.com/snehilvj/dash-mantine-components) have a great way of displaying each widget with examples and full docs, as shown [here](https://www.dash-mantine-components.com/components/checkbox), with one page per widget.

**Comment by bryevdv** [comment]
I don't think `autoapi` is going to be an option, AFAICT it is just completely independent of `autodoc` directives meaning all our custom extensions for models, properties, enums, etc. will not function. I will look at `sphinx-apidoc` and/or `sphinx.ext.autosummary` next, which should explicitly work together with `autodoc`.

**Comment by bryevdv** [comment]
@ianthomas23 I am actually starting to come to the conclusion that "one page per model" may in fact not be possible unless we adopt the fine-grained split you suggested... (or else write yet another custom extension or literally hand-author hundreds of ReST files, but I am loathe to consider either of those options)

**Comment by bryevdv** [comment]
unfortunately `autosummary` is also currently a non-starter due to a Sphinx limitation https://github.com/sphinx-doc/sphinx/issues/12021

Bokeh has unusual requirements, and I am not sure there is any immediate path forward without more custom tooling. I'm not sure when/if I would have bandwidth to look at that.

**Comment by peekxc** [comment]
Along a similar vein, and sort of foreshadowed in #12707, if the aim is to improve the docs one real issue to address to cut down the massive amount of information is not just one-page-per-class but also to remove redundant / inherited information. 

Take, for example, the RangeTool vs the TapTool docs. Here's [a diff of their docs](https://www.diffchecker.com/Odh4fp3P/): of the 465 lines of the line constituting TapTools docs, 431 of them are inherited boilerplate shared with RangeTool (and many others). 

In other words, if I want to read about what `TapTool` does and how it differs from `RangeTool` (say), I have to explicitly scan over / ignore 93% of the documentation to find the information I want. This is quite cumbersome! 

A single-page-per-class could help here, but actually one of the real culprits is the massive inheritance trees!

**Comment by bryevdv** [comment]
> cut down the massive amount of information is not just one-page-per-class but also to remove redundant / inherited information.

Just to be clear the reference docs did not originally duplicate inherited information. It was changed to the current form explicitly, and on purpose, because years ago, some users *strenuously, **vociferously*** complained that not having every entry be "complete" meant even calling it a "reference guide" was a horrible misleading lie. So. Truly, there is absolutely no satisfying everyone. 

Yes, I personally also think there is too much information. But there is *alot* of factoring in order to optimize the cross-runtime implementation, and not all of that factoring will be useful, relevant, or obvious to users. In other words, I think just deleting `:inherited-members:` and calling it a day will absolutely result in too *little* information on many entries, without question (even with links to base classes). For example: going to the `ColumnDataSouce` page and not finding an entry for `selection` (because it happens to be defined on the obscure `DataSource` base) would be bad. 

I guess offhand, I think:

* All *Bokeh properties* should be listed for every model class, regardless of inheritance.
* Most or all *methods* do not need to be repeated. Most users don't care about methods on models in the first place.

We'd probably have to do customization to our sphinx extensions to make this happen, and then turn `:inherited-members:`  off.

OTOH maybe this approach would itself be a confusing inconsistency. So, I don't really know what a good solution is here. 

Edit: though, thinking about it more, I guess I am not sure that just deleting `:inherited-members:` wouldn't, in fact, do just what I described above (the sphinx extensions may already ignore `:inherited-members:`, more or less) but regardless there is the matter of deciding whether this approach is the best one.

**Comment by peekxc** [comment]
> But there is alot of factoring in order to optimize ... we'd probably have to do customization to our sphinx extensions to make this happen

This sounds like maybe a limitation of the sphinx pipeline you're using. Maybe an issue should be raised somewhere else regarding this? 

> For example: going to the ColumnDataSouce page and not finding an entry for selection (because it happens to be defined on the obscure DataSource base) would be bad.

I guess this is a bit subjective. Though argument forwarding (à la `*args` and `**kwargs`) is somewhat common in Python, I admit it's not as ubiquitous as it is in e.g. R (or when it happens, it's only one level deep). I personally have no problems tracing up the inheritance hierarchy to understand the configurability of the API, though I admit the vast majority of the projects I work with (e.g. numpy & scipy) don't have _near_ the level of inheritance as bokeh. 

I do think inherited member should at the very least be treated differently. Even a simple fix like a collapsible (which is already done for the JSON prototypes) would do wonders, like: 


## TapTool 
- mode
- behavior
- gesture
- modifiers
- ...etc  
<details>
<summary> Inherited </summary>
- icon
- id 
- js_event_callbacks
- etc. 
</details>

**Comment by bryevdv** [comment]
> This sounds like maybe a limitation of the sphinx pipeline you're using. Maybe an issue should be raised somewhere else regarding this?

Not at all. The intent of leaving off `:inherited-members:` is to leave off inherited members. It does exactly what it says. My first comment is that we would need to do some work to circumvent the obvious stated behavior for our ununsual special case, and my second comment was that that may in fact already how our existing extensions behave. 

> I do think inherited member should at the very least be treated differently. Even a simple fix like a collapsible (which is already done for the JSON prototypes) would do wonders, like:

Putting things in a hidden, collapsible state makes linking directly a problem. I use direct anchor links in answers to support questions, if not daily, then at least several times a week. I am a big :-1: on this approach. 

I think we should not get off track here. Going to one page per class will be a huge improvement. That itself will also be somewhat of a lift to achieve, so it's all I plan to expend my own efforts on (when I have the time). Further discussion about other changes should go in a separate GitHub discussion or issue.
